npm/src/debug-service/debug-service.ts

```diff
@@ -30,18 +31,40 @@ export type IDebugServiceWorker = IDebugService & IServiceProxy;
 export class QSharpDebugService implements IDebugService {
   private debugService: DebugService;
 
+  // We need to keep a copy of the code for mapping diagnostics to utf16 offsets
+  private code: { [uri: string]: string } = {};
+
   constructor(wasm: QscWasm) {
     log.info("Constructing a QSharpDebugService instance");
     this.debugService = new wasm.DebugService();
   }
 
   async loadSource(path: string, source: string): Promise<boolean> {
+    this.code[path] = source;
     return this.debugService.load_source(path, source);
   }
 
-  async getStackFrames(): Promise<StackFrame[]> {
+  async getStackFrames(): Promise<IStackFrame[]> {
     const stack_frame_list = this.debugService.get_stack_frames();
-    const stack_frames: StackFrame[] = stack_frame_list.frames;
+
+    const stack_frames: IStackFrame[] = stack_frame_list.frames.map(
+      (frame: IStackFrame) => {
+        if (frame.path in this.code) {
+          const mappedSpan = mapUtf8UnitsToUtf16Units(
+            [frame.lo, frame.hi],
+            this.code[frame.path]
+          );
+          const result = {} as IStackFrame;
```

Comment:
A more canonical way to do lines 57 - 62 in JavaScript would be to use the object spread operator (https://stackoverflow.com/a/49491435). i.e. replace all those lines with `return {...frame, lo: mappedSpan[frame.lo], hi: mappedSpan[frame.hi]};`